### PR TITLE
refactor: move common steps from server and agent to prompkg

### PR DESCRIPTION
**Pull request by sthaha**
## Description

This commit partially address https://github.com/prometheus-operator/prometheus-operator/issues/5455 by refactoring server and agent operator code that updates the prometheus config secrets. It moves the common functionalities to pkg/prometheus (prompkg).

The refactor adds the following functions

Store Related
  * AddRemoteWritesToStore
  * AddRemoteReadToStore
  * AddAPIServerConfigToStore

Operator Related
  * MakeCompressedSecretForPrometheus

GH: https://github.com/prometheus-operator/prometheus-operator/issues/5455

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [x] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.) refactor

### CODE REVIEW

**Comment by sthaha** [comment]
I have currently marked the PR as a draft to ensure all tests pass, and that the maintainers can comment on the approach before adding more. 

cc: @simonpasquier

File: pkg/prometheus/store.go
Comment (sthaha) [inline pkg/prometheus/store.go:27]: Unit tests?

File: pkg/prometheus/statefulset.go
Comment (simonpasquier) [inline pkg/prometheus/statefulset.go:132]: (suggestion) I would 
* Merge `MakeConfigSecret` and `MakeConfigurationSecret` so the resulting function includes the compression of the raw config.
* get rid of `MakeEmptyConfigurationSecret`  as it should be equivalent to calling `MakeConfigurationSecret(p, c.config, nil)` 
```suggestion
func MakeConfigurationSecret(p monitoringv1.PrometheusInterface, config operator.Config, data []byte) (*v1.Secret, error) {
```

File: pkg/prometheus/store.go
Comment (simonpasquier) [inline pkg/prometheus/store.go:28]: I wouldn't do defensive programming. If store is nil, we have a bug and it's better to crash.

Comment (ArthurSens) [reply pkg/prometheus/store.go:28]: Hey, taking this opportunity to learn a bit more about the operator, what exactly are the stores? Why would it be a bug if they are nil?

Comment (sthaha) [reply pkg/prometheus/store.go:28]: How about we return an error rather than crash? I feel something odd about terminating a program abruptly, especially when we can handle that state gracefully ?

Comment (simonpasquier) [reply pkg/prometheus/store.go:28]: a nil store means that the developer forgot to pass a valid instance which is always a bug.

Comment (simonpasquier) [reply pkg/prometheus/store.go:28]: A store is a struct where we keep data referenced by secret or configmap keys such as TLS certificates, bearer tokens, .... If the same secret key is referenced by several resources, the operator will get it once from the Kube API. It also serves as a way to drop resources that reference non-existing data.

File: pkg/prometheus/statefulset.go
Comment (sthaha) [inline pkg/prometheus/statefulset.go:137]: @simonpasquier this part is different to the old code that I moved.

File: pkg/prometheus/store.go
Comment (ArthurSens) [reply pkg/prometheus/store.go:27]: While unit tests would be helpful for quick tests during development, we already have e2e tests that wouldn't pass if a regression happens here.

I wouldn't block the PR :)

**Review by ArthurSens — APPROVED** [review]

**Review by JoaoBraveCoding — APPROVED** [review]
lgtm

**Pull request merged**
